### ISSUE TITLE
[FR] [ST7920_Simulator.c] Add a character generator in order to use klipper

### ISSUE THREAD

**Issue by derBG** [bigtreetech/BIGTREETECH-TouchScreenFirmware#290]
Many users of your skr boards use klipper (https://github.com/KevinOConnor/klipper).

Please add a character generator to TFT / src / User / API / UI / ST7920_Simulator.c to make your tft24 screen and similiar usable.

Further information can be found in the Klipper issue at https://github.com/KevinOConnor/klipper/issues/2231

Your community will appreciate it very much

Thank you very much

**Comment by erik23de** [comment]
I'd like to expicitely agree to the post before. Please, add a character generator!

**Comment by DeviousPenguin** [comment]
I use Klipper, and an currently unable to use my TFT24 with it.

I looks like the emulation of the reprap display needs a character generator.

If you could add a character generator that would be really nice, as I'd like to use the TFT24 with the printer I am building.

**Comment by DeviousPenguin** [comment]
For anyone looking to do this, the ST7920 datasheet is a good resource:

> ST7920 provides character generation ROM supporting 8192 16 x 16 character fonts and 126 8 x 16 alphanumeric
characters. It is easy to support multi languages application such as Chinese and English. Two consecutive bytes are used to specify one 16x16 character or two 8x16 half-height characters. Character codes are written into DDRAM and the corresponding fonts are mapped from CGROM or HCGROM to the display drivers.

**Comment by kakou-fr** [comment]
yes please add it

**Comment by tpipits** [comment]
+1

**Comment by AnHardt** [comment]
The original font-ROM of the ST7920 is about 130 kByte.
Even for the larger processors of the displays, this is more than 50% of the flash memory.
Better do something more useful with the memory.

**Comment by JohnEdwa** [comment]
The ASCII half-height HCGROM is just 16K-bits, which is 2kB. The rest is for chinese characters, which I would assume are rarely needed with 3D printers.

**Comment by bigtreetech** [comment]
> The original font-ROM of the ST7920 is about 130 kByte.
> Even for the larger processors of the displays, this is more than 50% of the flash memory.
> Better do something more useful with the memory.

BTT's touch screen has an external flash(W25Qxx) storage bitmap fonts, so don't worry about this. This function should be easy to implement, but I'm busy recently, so I don't have time to test this. After a new version is released this week, I will start to implement this function

**Comment by combolek** [comment]
I would like to register another request to get this implemented. Thanks!

**Comment by mbgroot** [comment]
I'm very much looking forward to a solution to this problem. The printer is without screen :(

**Comment by kakou-fr** [comment]
no news ?

**Comment by dushyantahuja** [comment]
This would be great.

**Comment by TakeThisBanana** [comment]
I would also love to see the TFT35 E3 implemented.

**Comment by Lyngtoft** [comment]
Brilliant, just switched to klipper, would love to use my tft35

**Comment by bigtreetech** [comment]
![image](https://user-images.githubusercontent.com/38851044/81798875-196ad280-9543-11ea-981c-97dfc2d32681.png)

![image](https://user-images.githubusercontent.com/38851044/81798842-0fe16a80-9543-11ea-8e57-f750d2d5e24e.png)

Debuging... It will be realized soon

**Comment by mbgroot** [comment]
the font will depend on the klipper or is it hard to set in the screen firmware?

**Comment by JohnEdwa** [comment]
It's set by the screen, this whole thing is because Klipper uses the internal text generator on the display driver for the text, which was missing from this simulator. And as it is a simulator, it needs to include that character rom data as well which means it should be rather easy to swap it for a different font - if one exists or can be easily created, that is.

**Comment by jdlongenecker** [comment]
@bigtreetech thank you for working on this, I know a lot of people are excited to have some high quality display options!

**Comment by eboblin** [comment]
@bigtreetech any updates on this issue?

**Comment by TakeThisBanana** [comment]
> ![image](https://user-images.githubusercontent.com/38851044/81798875-196ad280-9543-11ea-981c-97dfc2d32681.png)
> 
> ![image](https://user-images.githubusercontent.com/38851044/81798842-0fe16a80-9543-11ea-8e57-f750d2d5e24e.png)
> 
> Debuging... It will be realized soon

How is the debugging comming along? I am really looking forward to using my tft35 e3 again :D

**Comment by JeffMedd** [comment]
I'll add a +1 of interest for this feature also! Thanks for prioritizing the development of this, it will be very much appreciated.

**Comment by maketo3D** [comment]
+1 very interested on it.

**Comment by RomRider** [comment]
Still a bunch of glitches, but this is great progress!!
![IMG_20200525_201843](https://user-images.githubusercontent.com/21064206/82835693-1964be00-9ec5-11ea-9076-6beb7a2f38b7.jpg)

**Comment by RomRider** [comment]
@bigtreetech, when the printer starts with the latest master branch version, I have to switch to touchmode and back to marlin mode before something is displayed on the screen else it stays black.
I have `SPI_ReEnable(0);` set + the other function updated and a TFT35 v3. Also I have a SKR1.3 board.

**Comment by TakeThisBanana** [comment]
> ![image](https://user-images.githubusercontent.com/38851044/81798875-196ad280-9543-11ea-981c-97dfc2d32681.png)
> 
> ![image](https://user-images.githubusercontent.com/38851044/81798842-0fe16a80-9543-11ea-8e57-f750d2d5e24e.png)
> 
> Debuging... It will be realized soon

How are things comming along?

**Comment by cskozlowski** [comment]
Any update on this front?  I have a tft35 display laying around that I would really like to use with klipper on my Voron 2.4 build.

**Comment by TakeThisBanana** [comment]
How is the status of developement?

**Comment by GrumpaV2** [comment]
Just bought TFT 70 in big summer sale to use with Klipper on SKR 1.4.... Please implement this in firmware update

**Comment by marvelous240** [comment]
Would also love to see this implemented soon!

**Comment by kakou-fr** [comment]
I tried on klipper with a TFT 24 : 
1) a lot of glitch
2) 90% of the time, when the print  boot, the display only show " .... simulator"
3) when the display work, after 1-2 hour, it stop working correctly and only display the bed icon on all the screen

**Comment by jariivanoff** [comment]
Adding a wish to get this fully implemented! :-D

**Comment by weaslus** [comment]
yeah, this is going to be my browser homepage for the foreseeable future...

**Comment by GrumpaV2** [comment]
I should have just bought Duet and Paneldue! What a waste of 65 bucks... Maybe I can hook it up to my kids Ender 3

**Comment by aklef** [comment]
@bigtreetech @RomRider bump! Devs, we love you. Do you need anything? Can I help? A glass of water perhaps? We just want y’all to keep us updated please!

**Comment by RomRider** [comment]
I'm not a dev here, it was just the result on my printer 😊

**Comment by oldman4U** [comment]
Could you please be so kind and close this ticket. Your feature request is covered in the pinned Feature Request ticket #1177 

Thank you

**Comment by RomRider** [comment]
This isn't in any way related to #1177

**Comment by oldman4U** [comment]
@RomRider 

Sorry typo #1170 not 1177.

Thank you

**Comment by RomRider** [comment]
But why would you want to close this FR since it is not fully implemented? I don't understand the logic?

**Comment by oldman4U** [comment]
Because there is now one centralized place instead of many tickets on a
page where nobody can find it.

Jérôme W. <notifications@github.com> schrieb am Mi. 21. Okt. 2020 um 00:05:

**Comment by RomRider** [comment]
That doesn't make any sense. Discussions on several features shouldn't happen in one mega thread it's going to be a massive mess!

Each feature should get its own discussion thread and then you can maintain your issues which references them all if you want...

**Comment by oldman4U** [comment]
Are you the owner of this ticket?

There is a link to this ticket for reference FR Ticket and because it is closed does not mean that it can not be used.

But if you are the owner and you prefer a ticket on page 2 or even further behind over a pinned ticket just let me know. Such tickets like inactive tickets will be removed from the FR Ticket in about one week.

**Comment by weaslus** [comment]
Are you the owner of this github?
What is your role here?

**Comment by oldman4U** [comment]
I am a user like you. 

Looking at other repositories like the SKR v1.3 I try to keep this one usable.

What are you doing to help the community to make this open source project a success?

**Comment by weaslus** [comment]
Im not goimg to mince words here. Appologies to those who dont care to recieve these strings of emails.

You seem like the self appointed sherrif. Your tone is tiresome. What happens if you leave the github due to boredom, busy work life, death? You want your way. Does your way really require that this feature request be closed, when it is still an active request?

**Comment by oldman4U** [comment]
I will hand it over - except I am death.

In case you want to do it let me know and we can do it right now.

Again. What are you doing? I can not develop so I try to help this way and
I am sure there is also something you could do.

This ticket can remain opened no need to close it. Are you the owner?


weaslus <notifications@github.com> schrieb am Mi. 21. Okt. 2020 um 09:46:

**Comment by marvelous240** [comment]
Shut up. All of you.

**Comment by oldman4U** [comment]
;-)

**Comment by kakou-fr** [comment]
I resolved my problems, I brought a Fysetc MINI 12864 :). It works and I can control thecolor of the screen/button  ...

**Comment by DeviousPenguin** [comment]
The Fysetc is a nice workaround :) I'm doing the same, but even so it would be nice to be able to use the BTT screen as a backup or 2nd printer, I've pretty much gone Klipper only.

**Comment by Stellarspace** [comment]
> ;-)

Considering it's unusual to close an issue before the problem has been fixed on GitHub, maybe you should leave a note somewhere suggesting that some issues that are closed haven't been fixed and are actually being worked on.

**Comment by oldman4U** [comment]
@Stellarspace #1170 collects all FR and gives them a prominent place, instead of being almost the last one on page 4. It is also the place where future development will be organised and developers can check what's needed and if someone is already working on it.

But the owner of this ticket does not respond, so this FR will be even less important in the near future.;-(

**Comment by RomRider** [comment]
You are reinventing github and how development workflows work. Good luck.
A product (open source or not) needs a Product Manager which defines a roadmap, sets priorities, etc... Apparently that's not what you do.

**Comment by JeremyKennedy** [comment]
Just because the owner of this ticket doesn't respond, does not mean this ticket is less important. I am still waiting on this ticket, and I'm subscribed to it, so that any progress on it gets sent to me via email. This whole mess you've started with "cleaning up" the issues repo is messing with that... and as I'm sure you've seen nearly all responses have been negative. As @RomRider said here and I said in #1170, you are reinventing development workflows which have worked for thousands of projects so far.

Just because a repository has a large list of open/old issues doesn't mean they need to be closed. It's very likely they are still issues, and maybe the issue owner doesn't care anymore, but they should still remain open.

**Comment by oldman4U** [comment]
Working in the production and broadcast industry for more than 25 years, half of the time as product manager, I am used to GIT, Atlassian Tools, on premise and cloud based support systems and so on.

Usually the owner of a repository is taking care of it, and makes decisions. Both is not the case here. An existing development workflow - I can not see it. Instead, I can see many developers leaving.

If you read my comments again you will see, that I never wrote that this ticket is less important than others - in fact, the opposite is the truth. I tried to give it a better chance to happen one day, by changing the way FR's are handled. But I almost get the feeling, that the status quo worked so well in the past, that changes are not welcome.

I can accept this.

**Comment by gkatev** [comment]
Both methods are useful.
A centralized thread tracking all the FRs and their progress or what-not is helpful.
However, the individual issues should not be closed; it leads to confusion and there's no good reason to do so.

**Comment by derBG** [comment]
Hello friends,

a ticket will be closed when its done. This is the normal behavior in Ticket Systems for software development. So i will close this ticket when the topic is completed.

Thanks
derBG

**Comment by oldman4U** [comment]
Hi derBG.

Thank you for joining.

Of course, it is your right to close your own ticket whenever you want.
Yes, it is completely uncommon to close unsolved tickets. This would especially be true for tickets in a system, which is actively monitored and maintained by the owner, and where no alternative system is available.
It is also common to close a ticket once it is not needed any more. Many unused tickets made this repository unusable.
A Bot has been installed which will close unused tickets, even they are important.
Because of this, I added a centralised BUG list in addition to the FR list, so important Bugs are also covered.
This particular feature request is the first entry of the FR list, at the moment.
Closing all tickets which are part of the FR (and BUG) list, would help to make the ticket system more user friendly and maintainable.
Some understood and wanted to help and closed their tickets, some not.

Like it is your right to decide what you want to do with your ticket, it is also my right to decide when I spend enough time for a certain thing. And for this ticket the time is now.

Kind regards and all the best with your FR. I hope you will get it asap!

**Comment by mbgroot** [comment]
Has the work on this ticket been stopped?

**Comment by DeviousPenguin** [comment]
@mbgroot I think this is still being work on (hopefully)

**Comment by tsndr** [comment]
I doubt it 😂

**Comment by simonvez** [comment]
I really hope it will work some time soon. :) I would love to use klipper again

**Comment by tsndr** [comment]
And I would like to use the BTT screen again, beacuse I'm not ditching Klipper because of a stupid screen. If they don't fix it I'll never buy BTT stuff again!

**Comment by bung69** [comment]
BTT just lost another sale of a TFT35 because this is not fixed

**Comment by ETE-Design** [comment]
@bigtreetech Could you pls. look at this feature and make the last work to make it work as intended? Thanks....

**Comment by teeminus** [comment]
After digging through the code and the ST7920 datasheet I think I found the solution to make the TFT35 and likely other BTT screens work with klipper. I fixed the issue in my fork and can confirm that it works for the TFT35v3: https://github.com/teeminus/BIGTREETECH-TouchScreenFirmware

I am not sure if I should create a MR for that as I hardcoded SPI mode 1 as suggested somewhere else.

I also created a stripped down version of the firmware that has only the ST7920 emulator left as I don't use the touch functionality at all. Feel free to check it out as well: https://github.com/teeminus/NoTouchScreenFirmware

**Comment by ETE-Design** [comment]
@teeminus Great work... Is there support for touch in Klipper? Else the "Stripped" down menu should mabye be an Env to define when compiling, mabye called "Klipper Mode"?

**Comment by teeminus** [comment]
@ETE-Design I don't know if klipper has touch support, but I would guess there isn't as "real" ST7920 based screens just have a rotary knob.

Technically one could create klipper build targets for the different screen models. But it seems that the spi mode auto detection is not working correctly, so that needs to be disabled in that mode as well.

**Comment by DeviousPenguin** [comment]
@teeminus thank you very much for your work on this. I would be more than happy simply having a text only ST7920 firmware, as I only use Klipper now and I personally prefer to have no touchscreen functionality on a 3d printer as I'm clumsy and have sometimes accentally cancelled long prints in the past!

I only have the TFT24 however, so I'll give it a go, right now the screen has been gathering dust for almost 1 year so it would be nice to use it for *something*.

**Comment by teeminus** [comment]
@DeviousPenguin As not all BTT screens have the required "ST7920_SPI" define set, I updated the readme of the minimalistic firmware so one can easily see if the screen is currently supported. I also uploaded precompiled binaries, which came out as byproducts of the compile test.

**Comment by cropduster** [comment]
> After digging through the code and the ST7920 datasheet I think I found the solution to make the TFT35 and likely other BTT screens work with klipper. I fixed the issue in my fork and can confirm that it works for the TFT35v3: https://github.com/teeminus/BIGTREETECH-TouchScreenFirmware
> 
> I am not sure if I should create a MR for that as I hardcoded SPI mode 1 as suggested somewhere else.
> 
> I also created a stripped down version of the firmware that has only the ST7920 emulator left as I don't use the touch functionality at all. Feel free to check it out as well: https://github.com/teeminus/NoTouchScreenFirmware

The touchscreen firmware is gone. Anyone knows where I can find it? I currently use the notouchfirmware.

**Comment by teeminus** [comment]
That was just my fork to investigate the issue. I deleted that fork as it turns out the "fix" I found was no real fix for the issue.

**Comment by cropduster** [comment]
> That was just my fork to investigate the issue. I deleted that fork as it turns out the "fix" I found was no real fix for the issue.

Is there any way to increase the resolution of the notouchscreenfirmware?

**Comment by teeminus** [comment]
What do you mean by resolution? The firmware uses the biggest pixel size possible. There is no fullscreen mode because this produces fractional pixel sizes.

**Comment by cropduster** [comment]
The TFT35 has a resolution of 480x320, but the emulation utilizes 128x64, so it is quite blocky. Not sure if this can be scaled up somehow or if this is technically impossible.

**Comment by teeminus** [comment]
The firmware uses a pixel size of 3x3 for the TFT35. The image will always be blocky because there is no edge smoothing or something like that.

**Comment by DaStivi** [comment]
whats the status right now of this?? i'm testing klipper currently, and with BTT Firmware i did have ah "working" screen sometimes, sometimes after i turn the decoder wheel, and sometimes its only displaying rubish... as far as i understand there happens same bit mismatch, does this also happen on the noscreenfirmware ?  @teeminus (i gonna test this in a seconds on my B1 TFT35)

**Comment by teeminus** [comment]
It depends on your printer mainboard. On some mainboards, the display is not reset when Klipper is restarted which causes the bit shifts. This problem is independent from the TFT firmware.

However, for Klipper there is a "emulated_st7920" display driver which has been developed to fix this problem with the NoTouchFW. This might fix the communication problems for the BTT firmware as well.

**Comment by DaStivi** [comment]
@teeminus i've tested the notouchFW... looks like it works better with klipper as the current-normal BTT TFT... as i didn't get any character gitter anymore....  BUT i can't touch the wheel.... ah couple of times i had  klipper to "shutdown", looks like i fixed this too now... as something odd with PI + picam v2 goes on... i even had this before with marlin and octoprint, but fixed it with wifi firmware downgrade... did now an update with kiauh and looks like i updated the firmware again, but can't revert it back to the old state now... idk... but anyway when webcam is running pi's cpu usage gets high and wifi starts dropping packages... and i think i had some random shutdowns from klipper because of this...  so not sure if this was ah coincidence when i pressed the encoder, but for sure i can't get into the menu...  also tried so set the klipper config allready to the emulated_st9720

**Comment by teeminus** [comment]
@DaStivi The encoder when and button is handled by Klipper, not by the TFT firmware.

**Comment by DaStivi** [comment]
Thx for explanation, indeed it worked once yesterday... So sometimes the screen still looks like to freeze, aber had also ah garbeled screen yesterday again.. I'd might install your debug Version... But maybe also try again btt FW with the emulated screen Config again...

**Comment by DaStivi** [comment]
@teeminus flashed the debug today as i got random freezes of the tft.... started another calibration print, and noticed the new debug ui... allready noticed some lag of the display... but after that it resumed... look at the printer 10min later and just welcomed with following garbage... as you may can see the debug counter still counts as expected i guess...


![image](https://user-images.githubusercontent.com/7464802/115278594-ed3dd580-a145-11eb-98ed-f786b69517d3.png)

**Comment by teeminus** [comment]
@DaStivi Could you please create a new issue at the NoTouchFW repo? I think it's better to discuss you problem there as it is not related to this firmware.

**Comment by Vybramo** [comment]
+1

**Comment by oldman4U** [comment]
Good news for all Klipper users. User terminus was able to solve an outstanding issue together with BTT and now a PR is available.

#1913

Happy testing
